flink-clients/src/test/java/org/apache/flink/client/deployment/application/ApplicationDispatcherBootstrapTest.java

```diff
@@ -301,69 +306,132 @@ public void testApplicationIsStoppedWhenStoppingBootstrap() throws Exception {
 
 		// verify that the application task is being cancelled
 		assertThat(applicationExecutionFuture.isCancelled(), is(true));
 	}
 
 	@Test
-	public void testClusterShutdownWhenStoppingBootstrap() throws Exception {
+	public void testErrorHandlerIsCalledWhenStoppingBootstrap() throws Exception {
+		final TestingDispatcherGateway.Builder dispatcherBuilder = new TestingDispatcherGateway.Builder()
+				.setSubmitFunction(jobGraph -> CompletableFuture.completedFuture(Acknowledge.get()))
+				.setRequestJobStatusFunction(jobId -> CompletableFuture.completedFuture(JobStatus.RUNNING));
+
+		// we're "listening" on this to be completed to verify that the error handler is called.
+		// In production, this will shut down the cluster with an exception.
+		final CompletableFuture<Void> errorHandlerFuture = new CompletableFuture<>();
+		final ApplicationDispatcherBootstrap bootstrap = createApplicationDispatcherBootstrap(
+				2, dispatcherBuilder.build(), scheduledExecutor, errorHandlerFuture::completeExceptionally);
+
+		final CompletableFuture<Acknowledge> shutdownFuture = bootstrap.getClusterShutdownFuture();
+
+		bootstrap.stop();
+
+		// we call the error handler
+		assertException(errorHandlerFuture, CancellationException.class);
+
+		// we return a future that is completed exceptionally
+		assertException(shutdownFuture, CancellationException.class);
+	}
+
+	@Test
+	public void testErrorHandlerIsCalledWhenSubmissionFails() throws Exception {
+		final Tuple1<Boolean> clusterShutdown = new Tuple1<>(false);
+		final TestingDispatcherGateway.Builder dispatcherBuilder = new TestingDispatcherGateway.Builder()
+				.setSubmitFunction(jobGraph -> {
+					throw new FlinkRuntimeException("Nope!");
+				})
+				.setClusterShutdownFunction(status -> {
+					if (clusterShutdown.f0) {
+						throw new FlinkRuntimeException("This should be called only once");
+					}
+					clusterShutdown.f0 = true;
+					return CompletableFuture.completedFuture(Acknowledge.get());
+				});
+
+		// we're "listening" on this to be completed to verify that the error handler is called.
+		// In production, this will shut down the cluster with an exception.
+		final CompletableFuture<Void> errorHandlerFuture = new CompletableFuture<>();
+		final TestingDispatcherGateway dispatcherGateway = dispatcherBuilder.build();
+		final ApplicationDispatcherBootstrap bootstrap = createApplicationDispatcherBootstrap(
+				3, dispatcherGateway, scheduledExecutor, errorHandlerFuture::completeExceptionally);
+
+		final CompletableFuture<Acknowledge> shutdownFuture = bootstrap.getClusterShutdownFuture();
+
+		// we call the error handler
+		assertException(errorHandlerFuture, ApplicationExecutionException.class);
+
+		// we return a future that is completed exceptionally
+		assertException(shutdownFuture, ApplicationExecutionException.class);
+
+		assertFalse(clusterShutdown.f0);
+		dispatcherGateway.shutDownCluster(ApplicationStatus.SUCCEEDED);
+		assertTrue(clusterShutdown.f0);
```

Comment:
Why is this necessary? Couldn't we write the test w/o it and then directly fail if `shutDownCluster` is being called?

Reply:
Or we simply add a counter to `setClusterShutdownFunction` and assert that its value `0` at the end of the test?